src/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestService.cs

```diff
@@ -29,66 +33,99 @@ public TestService(string serviceName)
+            this.Server = new NamedPipeServerStream(serviceName);
+            serverStarted = this.Server.WaitForConnectionAsync();
         }
 
+        private NamedPipeServerStream Server { get; }
+
         protected override void OnContinue()
         {
-            WriteLog(nameof(OnContinue));
             base.OnContinue();
+            var task = Task.Run(() => WriteStreamAsync(PipeMessageByteCode.Continue));
+            task.Wait();
         }
 
         protected override void OnCustomCommand(int command)
         {
-            WriteLog(nameof(OnCustomCommand) + " command=" + command);
             base.OnCustomCommand(command);
+            var task = Task.Run(() => WriteStreamAsync(PipeMessageByteCode.OnCustomCommand, command));
+            task.Wait();
         }
 
         protected override void OnPause()
         {
-            WriteLog(nameof(OnPause));
             base.OnPause();
+            var task = Task.Run(() => WriteStreamAsync(PipeMessageByteCode.Pause));
+            task.Wait();
         }
 
         protected override void OnSessionChange(SessionChangeDescription changeDescription)
         {
-            WriteLog(nameof(OnSessionChange) + " change=" + changeDescription.ToString());
             base.OnSessionChange(changeDescription);
         }
 
         protected override bool OnPowerEvent(PowerBroadcastStatus powerStatus)
         {
-            WriteLog(nameof(OnPowerEvent) + " status=" + powerStatus.ToString());
             return base.OnPowerEvent(powerStatus);
         }
 
         protected override void OnShutdown()
         {
-            WriteLog(nameof(OnShutdown));
             base.OnShutdown();
         }
 
         protected override void OnStart(string[] args)
         {
-            File.Delete(GetLogPath(ServiceName));
-
-            WriteLog(nameof(OnStart) + " args=" + string.Join(",", args));
             base.OnStart(args);
+            if (args.Length == 4 && args[0] == "StartWithArguments")
+            {
+                Debug.Assert(args[1] == "a");
+                Debug.Assert(args[2] == "b");
+                Debug.Assert(args[3] == "c");
+                var task = Task.Run(() => WriteStreamAsync(PipeMessageByteCode.Start));
+                task.Wait();
+            }
         }
 
         protected override void OnStop()
         {
-            WriteLog(nameof(OnStop));
             base.OnStop();
+            var task = Task.Run(() => WriteStreamAsync(PipeMessageByteCode.Stop));
+            task.Wait();
+        }
+
+        private async Task WriteStreamAsync(PipeMessageByteCode code, int command = 0)
+        {
+            Task writeCompleted;
+            if (serverStarted.IsCompleted)
+            {
+                if (code == PipeMessageByteCode.OnCustomCommand)
+                {
+                    writeCompleted = Server.WriteAsync(new byte[] { (byte)command }, 0, 1);
+                    await writeCompleted.TimeoutAfter(60000);
```

Comment:
>; [](start = 60, length = 1)

nit: good practice: append ConfigureAwait(false) unless you need to preserve the thread context on continuation, which you typically need only in special cases (e.g. UI thread).